wardrobe/tradable_kitties/src/lib.rs

```diff
@@ -0,0 +1,573 @@
+            }
+            ConstraintCheckerError::NewChildDnaIncorrect => {
+                TradableKittyConstraintCheckerError::NewChildDnaIncorrect
+            }
+            ConstraintCheckerError::NewChildFreeBreedingsIncorrect => {
+                TradableKittyConstraintCheckerError::NewChildFreeBreedingsIncorrect
+            }
+            ConstraintCheckerError::NewChildHasNonZeroBreedings => {
+                TradableKittyConstraintCheckerError::NewChildHasNonZeroBreedings
+            }
+            ConstraintCheckerError::NewChildIncorrectParentInfo => {
+                TradableKittyConstraintCheckerError::NewChildIncorrectParentInfo
+            }
+            ConstraintCheckerError::TooManyBreedingsForKitty => {
+                TradableKittyConstraintCheckerError::TooManyBreedingsForKitty
+            }
+            ConstraintCheckerError::NotEnoughFreeBreedings => {
+                TradableKittyConstraintCheckerError::NotEnoughFreeBreedings
+            }
+            ConstraintCheckerError::MintingNothing => {
+                TradableKittyConstraintCheckerError::MintingNothing
+            }
+            ConstraintCheckerError::MintingWithInputs => {
+                TradableKittyConstraintCheckerError::MintingWithInputs
+            }
+        }
+    }
+}
+
+#[derive(
+    Serialize,
+    Deserialize,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Clone,
+    Encode,
+    Decode,
+    Hash,
+    Debug,
+    TypeInfo,
+)]
+pub enum TradableKittyConstraintChecker<const ID: u8> {
+    /// A mint transaction that creates Tradable Kitties
+    Mint,
+    /// A typical Breed transaction where kitties are consumed and new family(Parents(mom,dad) and child) is created.
+    Breed,
+    ///Update various properties of kitty.
+    UpdateProperties,
+    ///Can buy a new kitty from others
+    Buy,
+}
+
+pub trait Breed {
+    /// The Cost to breed a kitty.
+    const COST: u128;
+    type Error: Into<TradableKittyConstraintCheckerError>;
+    fn can_breed(mom: &TradableKittyData, dad: &TradableKittyData) -> Result<(), Self::Error>;
+    fn check_new_family(
+        mom: &TradableKittyData,
+        dad: &TradableKittyData,
+        new_family: &[DynamicallyTypedData],
+    ) -> Result<(), Self::Error>;
+}
+
+trait UpdateKittyProperty {
+    /// The Cost to update a kitty property if it is not free.
+    const COST: u128;
+    /// Error type for all Kitty errors.
+    type Error: Into<TradableKittyConstraintCheckerError>;
+
+    fn check_updated_kitty(
+        original_kitty: &TradableKittyData,
+        updated_kitty: &TradableKittyData,
+    ) -> Result<(), Self::Error>;
+}
+
+trait Buy {
+    /// Error type for all Kitty errors.
+    type Error: Into<TradableKittyConstraintCheckerError>;
+
+    fn can_buy(
+        input_data: &[DynamicallyTypedData],
+        output_data: &[DynamicallyTypedData],
+    ) -> Result<(), Self::Error>;
+}
```

Comment:
Trait and functions need docs. Is the trait necessary? Do you imagine some other type might implement this trait? If not, just implement the method directly on the type.

Reply:
Did you ever implement this trait? I'm not seeing your buy logic anywhere.